Directly create zarr datacubes on Cube.create and Cube.update

meggart opened the pull request:
This is my first working implementation for this change. From what I tested so far, the resulting data is the same as in the NetCDF cubes in the old generation mode. 

This is still WIP, I will do a bit code cleanup tomorrow and fix the unit tests. However, all comments are most welcome.

meggart committed:
initial implementation of directly genrating zarr cubes

meggart commented:
In particular, since python is not my goto-language, I have no idea about variable naming conventions, PEP8, etc, so probably did a lot of things wrong also on this side, any suggestions here would be great.

meggart committed:
actually write dimension attributes and close provider when finished

meggart committed:
make tests pass

meggart commented:
I think all tests related to this PR are pass now. The failing tests in test_aggregation are already broken on master, so I don't know what is going on there.

forman commented on README.rst:
```diff
@@ -113,6 +113,7 @@ Production and test code dependencies
 * netCDF4  >= 1.2
 * numpy >= 1.9
 * gridtools
+* zarr
```
Please specify version

forman commented on esdl/cube.py:
```diff
@@ -75,8 +78,15 @@ def open(base_dir):
             raise IOError('data cube base directory does not exists: %s' % base_dir)
-        config = CubeConfig.load(os.path.join(base_dir, 'cube.config'))
-        return Cube(base_dir, config)
+        zg = zarr.open_group(base_dir)
```
Please leave a code comment to indicate what's going on in the following. Or better extract a private method with a name indicating what it does and a comment why it is needed.

meggart replied:
I have re-written this part so that cube attributes are written through xarray and not through pure zarr, which should improve readability of this part

forman commented on esdl/cube.py:
```diff
@@ -91,22 +101,127 @@ def create(base_dir, config=CubeConfig()):
+        dsz = ds.to_zarr(base_dir)
+    
+
+        # Write group attributes
```
The following is just metadata attribute setting. Why don't you modify the `ds.attrs` directly?

forman commented on esdl/cube.py:
```diff
@@ -91,22 +101,127 @@ def create(base_dir, config=CubeConfig()):
+        
         return Cube(base_dir, config)
 
-    def close(self):
```
Keep `close()` for backward compatibility. Document it as deprecated and no-op.

forman commented on esdl/cube.py:
```diff
@@ -91,22 +101,127 @@ def create(base_dir, config=CubeConfig()):
-        self._closed = True
 
-    def update(self, provider: 'CubeSourceProvider'):
+    def update(self, provider: 'CubeSourceProvider', n_imagecache = 10):
```
Please rename `n_imagecache` to `image_cache_size`

forman commented on esdl/cube.py:
```diff
@@ -147,160 +274,89 @@ def update(self, provider: 'CubeSourceProvider'):
+        #First we unpack the data into 3d-numpy arrays
+        for var_name in varnames:
+            ds = datasets[var_name]
+            i0   = imagecache[0][0]
```
PEP8: Do not align `=`.

forman commented on esdl/cube.py:
```diff
@@ -147,160 +274,89 @@ def update(self, provider: 'CubeSourceProvider'):
+            im_first = imagecache[0][1][var_name]
+            image_all = np.full((iend-i0, im_first.shape[0], im_first.shape[1]),ds.fill_value,ds.dtype)
+            for entry in imagecache:
+                i1     = entry[0]
```
PEP8: Do not align `=`.

forman reviewed (changes requested):
Hi Fabian, 
Thanks very much for this PR! 
Before I continue my review, I'd like to discuss the general approach with you. What about tomorrow or Thursday?
Regards
    Norman

meggart commented:
This would be great. I am available all day tomorrow.

On 16.04.19 14:48, Norman Fomferra wrote:
>
> *@forman* requested changes on this pull request.
>
> Hi Fabian,
> Thanks very much for this PR!
> Before I continue my review, I'd like to discuss the general approach 
> with you. What about tomorrow or Thursday?
> Regards
> Norman
>
> ------------------------------------------------------------------------
>
> In README.rst 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275771158>:
>
> > @@ -113,6 +113,7 @@ Production and test code dependencies
>   * netCDF4  >= 1.2
>   * numpy >= 1.9
>   * gridtools
> +* zarr
>
> Please specify version
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275771817>:
>
> > @@ -75,8 +78,15 @@ def open(base_dir):
>   
>           if not os.path.exists(base_dir):
>               raise IOError('data cube base directory does not exists: %s' % base_dir)
> -        config = CubeConfig.load(os.path.join(base_dir, 'cube.config'))
> -        return Cube(base_dir, config)
> +        zg = zarr.open_group(base_dir)
>
> Please leave a code comment to indicate what's going on in the 
> following. Or better extract a private method with a name indicating 
> what it does and a comment why it is needed.
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275775630>:
>
> > +        lat_dim  = ('lat', lat_vals)
> +        lon_dim  = ('lon', lon_vals)
> +
> +        time_bnds_var = xr.DataArray(time_bnds_vals,coords = {'time': time_dim, 'bnds': bndsdim}, dims=['time', 'bnds'])
> +        lat_bnds_var = xr.DataArray(lat_bnds_vals,coords = {'lat': lat_dim, 'bnds': bndsdim}, dims=['lat', 'bnds'])
> +        lon_bnds_var = xr.DataArray(lon_bnds_vals,coords = {'lon': lon_dim, 'bnds': bndsdim}, dims=['lon', 'bnds'])
> +
> +        ds = xr.Dataset({'time_bnds': time_bnds_var,
> +                         'lat_bnds' : lat_bnds_var,
> +                         'lon_bnds' : lon_bnds_var,
> +                        })
> +
> +        dsz = ds.to_zarr(base_dir)
> +
> +
> +        # Write group attributes
>
> The following is just metadata attribute setting. Why don't you modify 
> the |ds.attrs| directly?
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275776198>:
>
> >          return Cube(base_dir, config)
>   
> -    def close(self):
>
> Keep |close()| for backward compatibility. Document it as deprecated 
> and no-op.
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275776859>:
>
> >  
> -    def update(self, provider: 'CubeSourceProvider'):
> +    def update(self, provider: 'CubeSourceProvider', n_imagecache = 10):
>
> Please rename |n_imagecache| to |image_cache_size|
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275777065>:
>
> > -        time_bnds = dataset.variables['time_bnds']
> -        if time_bnds[time_index, 1] != t2:
> -            print("Warning: Time stamps discrepancy: %f is is not %f" % (time_bnds[time_index, 1], t2))
> -            print("target start: %s, target end %s" % (target_start_time, target_end_time))
> -
> -        var_variable = dataset.variables[var_name]
> -        var_variable[time_index, :, :] = image
> -
> -    def _init_variable_dataset(self, provider, dataset, variable_name, start_year):
> +
> +
> +    def _write_images(self, provider, datasets, imagecache, varnames, n_imagecache):
> +        #First we unpack the data into 3d-numpy arrays
> +        for var_name in varnames:
> +            ds = datasets[var_name]
> +            i0   = imagecache[0][0]
>
> PEP8: Do not align |=|.
>
> ------------------------------------------------------------------------
>
> In esdl/cube.py 
> <https://github.com/esa-esdl/esdl-core/pull/76#discussion_r275777109>:
>
> > -        var_variable = dataset.variables[var_name]
> -        var_variable[time_index, :, :] = image
> -
> -    def _init_variable_dataset(self, provider, dataset, variable_name, start_year):
> +
> +
> +    def _write_images(self, provider, datasets, imagecache, varnames, n_imagecache):
> +        #First we unpack the data into 3d-numpy arrays
> +        for var_name in varnames:
> +            ds = datasets[var_name]
> +            i0   = imagecache[0][0]
> +            iend = imagecache[-1][0] + 1
> +            im_first = imagecache[0][1][var_name]
> +            image_all = np.full((iend-i0, im_first.shape[0], im_first.shape[1]),ds.fill_value,ds.dtype)
> +            for entry in imagecache:
> +                i1     = entry[0]
>
> PEP8: Do not align |=|.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/esa-esdl/esdl-core/pull/76#pullrequestreview-227160698>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ACbAKzxcMQSSBRmcoqBNbUPtggq_uGPqks5vhcabgaJpZM4clRri>.
>

meggart committed:
apply pep8

meggart committed:
Move attribute writing to xarray

meggart committed:
let user specify bounding box

Commit:
Remove check for lon and lat values

I want to build regional ESDL datacubes with data which is in a regional
projection like for example UTM.
It seems to work, when I remove the lat and lon value checks, because
these are only applicable in EPSG:4326.
This change needs the boundingbox branch.

dzelge commented:
@meggart In this PR on the other hand several tests fail. Do they run locally on your machine

dzelge committed:
- Changed _str_ to a non-html representation
- Renamed CubesStore to CubeStore
- Added a legacy CubesStore class with a depracation warning

meggart commented:
I finally had some time to look at the tests again. One main issue is that currently, with my rewrites, you can not start or end a cube in the middle of a year. I would be required for a cube to start on a first of January and end on a December 31st. 

My question would be: do we need this feature? If yes I could implement it, but it might take a bit of time. Note that this is only about metadata. It is still possible for a provider to only write the data that it provides and the remaining chunks just stay uninitialized.

meggart commented:
Are there any opinions if we need to support cubes that start/end in the middle of a year?

dzelge commented:
I think there is no need to specifically support that. The current setup can handle such data but will, of course, have half a year of Nones. However, when we keep the data pruned, it should not matter that much.

 

 

 

------------------------

Helge

Brockmann Consult

 

Von: Fabian Gans <notifications@github.com> 
Gesendet: Dienstag, 24. September 2019 09:47
An: esa-esdl/esdl-core <esdl-core@noreply.github.com>
Cc: Helge Dzierzon <helge.dzierzon@brockmann-consult.de>; Review requested <review_requested@noreply.github.com>
Betreff: Re: [esa-esdl/esdl-core] RFC: Directly create zarr datacubes on Cube.create and Cube.update (#76)

 

Are there any opinions if we need to support cubes that start/end in the middle of a year?

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/esa-esdl/esdl-core/pull/76?email_source=notifications&email_token=AKDZYIJN4TAAHS6WX3UHHI3QLHAY5A5CNFSM4HEVDLRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7NNS5A#issuecomment-534436212> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AKDZYIPSHOBJV4SYOQ5ZDBTQLHAY5ANCNFSM4HEVDLRA> .  <https://github.com/notifications/beacon/AKDZYIOHJMPVBINWKIA6YVLQLHAY5A5CNFSM4HEVDLRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7NNS5A.gif>

forman commented:
@meggart don't know if this is clear, by *pruned* @dzelge refers to the `xcube prune` tool, that deletes all data chunk files that only contain a variable's `fill_value`. So, even if a one variable starts in the middle of a yeat and others at the start, there will be no extra space needed for the fill value chunks.

meggart committed:
Remove cube_access interface. This is not needed anymore since we can simply use zarr+xarray, which is well-documented and battle-tested

meggart committed:
Fix unit tests

meggart committed:
PEP8

meggart committed:
rename n_imagecache

meggart commented:
I have finally updated the tests, applied PEP8 and implemented the change requests by @forman as good as I could. I also did a rebase so that no merge conflicts are there anymore. From my side this might be merged. @felixcremer I un-commented the grid validation code again so that this might be a problem for you again. Could you check? I think the only problem might be that we require ascending longitude and descending latitudes, but we might simply remove this check (see this https://github.com/esa-esdl/esdl-core/blob/zarrcreate/esdl/cube_config.py#L234-L236 )

meggart committed:
add minimum zarr version to readme

dzelge committed:
Merge remote-tracking branch 'origin/master'

dzelge committed:
Merge branch 'master' into zarrcreate

# Conflicts:
#	esdl/cube_store.py
